src/log-query/src/log_query.rs

```diff
@@ -0,0 +1,322 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chrono::{DateTime, Datelike, Duration, NaiveDate, NaiveTime, TimeZone, Utc};
+use table::table_name::TableName;
+
+use crate::error::{
+    EndBeforeStartSnafu, InvalidDateFormatSnafu, InvalidSpanFormatSnafu, InvalidTimeFilterSnafu,
+    Result,
+};
+
+/// GreptimeDB's log query request.
+pub struct LogQuery {
+    /// A fully qualified table name to query logs from.
+    pub table_name: TableName,
+    /// Specifies the time range for the log query. See [`TimeFilter`] for more details.
+    pub time_filter: TimeFilter,
+    /// Columns with filters to query.
+    pub columns: Vec<ColumnFilters>,
+    /// Maximum number of logs to return. If not provided, it will return all matched logs.
+    pub limit: Option<usize>,
+    /// Adjacent lines to return.
+    pub context: Context,
+}
+
+/// Represents a time range for log query.
+///
+/// This struct allows various formats to express a time range from the user side
+/// for best flexibility:
+/// - Only `start` is provided: the `start` string can be any valid "date" or vaguer
+///     content. For example: "2024-12-01", "2024-12", "2024", etc. It will be treated
+///     as an time range corresponding to the provided date. E.g., "2024-12-01" refers
+///     to the entire 24 hours in that day. In this case, the `start` field cannot be a
+///     timestamp (like "2024-12-01T12:00:00Z").
+/// - Both `start` and `end` are provided: the `start` and `end` strings can be either
+///     a date or a timestamp. The `end` field is exclusive (`[start, end)`). When
+///     `start` is a date it implies the start of the day, and when `end` is a date it
+///     implies the end of the day.
+/// - `span` with `start` OR `end`: the `span` string can be any valid "interval"
+///     For example: "1024s", "1 week", "1 month", etc. The `span` field is applied to
+///     the `start` or `end` field to calculate the other one correspondingly. If `start`
+///     is provided, `end` is calculated as `start + span` and vice versa.
+/// - Only `span` is provided: the `span` string can be any valid "interval" as mentioned
+///     above. In this case, the current time (on the server side) is considered as the `end`.
+/// - All fields are provided: in this case, the `start` and `end` fields are considered
+///     with higher priority, and the `span` field is ignored.
+///
+/// This struct doesn't require a timezone to be presented. When the timezone is not
+/// provided, it will fill the default timezone with the same rules akin to other queries.
+#[derive(Debug, Clone)]
+pub struct TimeFilter {
```

Comment:
Can support timestamp milliseconds number？

Reply:
Timestamp format can be millisecond or higher resolutions